### DISCUSSION TITLE
Problem about vips_text() font and italic

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2185] [Q&A]
Hello!
When I use vips_text() to create picture, different fonts get the same picture, sometime. Do i need to refresh something?
For example, the MicrosoftYaHei and the MicrosoftYaHei-Bold create the same picture, like this
![MicrosoftYaHei](https://user-images.githubusercontent.com/27042311/113692717-667a0900-9700-11eb-9d39-93750849150c.png)
![MicrosoftYaHei-Bold](https://user-images.githubusercontent.com/27042311/113692724-6974f980-9700-11eb-8ca7-59af6524a1d1.png)

And when I need italic text, it like this:
![example](https://user-images.githubusercontent.com/27042311/113691729-56155e80-96ff-11eb-91f3-77409567bee6.png)
The last letter is incomplete.
How can I solve these.

**Comment by jcupitt** [comment]
Hi @rebornfz,

Could you post a complete program that shows this problem? Also, what platform are you using, what version of libvips, and how did you build the library.

**Reply by rebornfz** [reply]
vips --version
vips-8.10.6-Tue Mar 23 20:52:58 UTC 2021

platform：
centos7， gcc8.3.0

build from release v8.10.6

example code and font file in this
https://drive.google.com/file/d/1moJ0L1Fb4KQEa8krvWKLDowDMj4XHKCZ/view?usp=sharing

**Reply by rebornfz** [reply]
This is my results.
[pic.zip](https://github.com/libvips/libvips/files/6264192/pic.zip)

**Reply by rebornfz** [reply]
Is it because Linux does not support these fonts well?

**Reply by rebornfz** [reply]
@jcupitt 
Are there other ways to render text images quickly and correctly?

**Comment by jcupitt** [comment]
I think you're setting the font names incorrectly, so it's giving you a fallback font.

Try this:

```
$ vips text x.png "Innovation in china 中国制造" --fontfile MicrosoftYaHei-Bold.ttf --font "YaHei Bold 72"
$ vips text x2.png "Innovation in china 中国制造" --fontfile MicrosoftYaHei.ttf --font "YaHei 72"
```

I see:

![x](https://user-images.githubusercontent.com/580843/114032471-3d7a8500-9874-11eb-95a9-b34ff88245fe.png)
![x2](https://user-images.githubusercontent.com/580843/114032485-423f3900-9874-11eb-81dc-a1676dc0d890.png)

**Comment by jcupitt** [comment]
On ubuntu, you can use `gnome-font-viewer` to see inside a font file, for example:

```
$ gnome-font-viewer "PingFang Medium.ttf"
```

Displays:

![x](https://user-images.githubusercontent.com/580843/114033514-40c24080-9875-11eb-8553-a03634bd4a0b.png)

So the font is called "PingFang SC" and it has a variant called "Medium". You'd need to call it something like "PingFang SC Medium 72" to render it.

**Reply by rebornfz** [reply]
I do this work on the server. It may be a problem with centos
This is mine
CentOS Linux release 7.5.1804 (Core) 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.5.1804 (Core) 
CentOS Linux release 7.5.1804 (Core)

**Reply by rebornfz** [reply]
This is too strange, the result of using the command line is not the same as the result of using the code

**Comment by jcupitt** [comment]
It'll be the same in code. I think you are not setting the font name correctly.

Try:

```
$ python3
Python 3.8.6 (default, Jan 27 2021, 15:42:20) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> a = pyvips.Image.text("Innovation in china 中国制造", fontfile="MicrosoftYaHei-Bold.ttf", font="YaHei Bold 72")
>>> a.write_to_file("x.png")
>>> b = pyvips.Image.text("Innovation in china 中国制造", fontfile="MicrosoftYaHei.ttf", font="YaHei 72")
>>> b.write_to_file("x2.png")
>>> 
```

You get the same result.

**Reply by rebornfz** [reply]
I get the name using the "fc-list", Is it right?

**Comment by jcupitt** [comment]
I think `fc-list` will only work for fonts that you've installed. It won't work for fonts you load with `fontfile`.

If you've installed fonts by copying files to `/usr/share/fonts`, I think you'll also need to run `fc-cache` to update (I think).

**Reply by rebornfz** [reply]
Yes, I am not setting the font name correctly. How did you get the name “YaHei Bold”?

**Reply by hanssonrickard** [reply]
I think he used "gnome font viewer" as described above, to load the font, there you can see the font name (YaHei) and the variants (Bold).

**Reply by rebornfz** [reply]
Thank you, The name obtained by fc-list may be wrong
